enarx-keep-sgx-shim/src/handler.rs

```diff
@@ -182,4 +191,70 @@ impl<'a> Handler<'a> {
             ret
         }
     }
+
+    /// Do a readv() syscall
+    pub fn readv(&mut self) -> u64 {
+        let fd = self.aex.gpr.rdi;
+        let trusted = unsafe {
+            from_raw_parts_mut(self.aex.gpr.rsi as *mut Iovec, self.aex.gpr.rdx as usize)
+        };
+
+        // Add up total size of buffers.
+        let bufsize = trusted.iter().fold(0, |a, e| a + e.size);
+
+        // Allocate some unencrypted memory.
+        let size = bufsize + size_of::<Iovec>() * trusted.len();
+        let map = match self.ualloc(size as u64) {
+            Err(errno) => return errno.into_syscall(),
+            Ok(map) => unsafe { from_raw_parts_mut(map, size as usize) },
+        };
+
+        // Split allocated memory into that used by the iovec struct array and that used by its buffers.
+        let (uiovec, ubuffer) = map.split_at_mut(size_of::<Iovec>() * trusted.len());
+
+        // Convert the prefix from a byte slice into an iovec slice.
+        let (pre, untrusted, suf) = unsafe { uiovec.align_to_mut::<Iovec>() };
+        if pre != [] || suf != [] {
+            self.attacked();
+        }
+
+        // Set pointers in unencrypted iovec slice to use the rest of the allocated memory.
+        let mut chunk = ubuffer.as_ptr();
+        for (t, mut u) in trusted.iter_mut().zip(untrusted.iter_mut()) {
+            u.base = chunk as *mut u8;
+            u.size = t.size;
+            chunk = unsafe { chunk.add(t.size) };
+        }
```

Comment:
Instead of converting the `ubuffer` to a pointer and then doing pointer math, instead track the offset and convert a subslice to a pointer. That will remove the untrusted code block.